url4fdf79ae2c_beckchr_juel_tgz-pJ8-de_odysseus_el_samples_xml_sax_XMELFilterJ8/src/javax/el/BeanELResolver.java

```diff
@@ -0,0 +1,666 @@
+	 *             if no suitable method can be found.
+	 * @throws ELException
+	 *             if an exception was thrown while performing (base, method) resolution. The thrown
+	 *             exception must be included as the cause property of this exception, if available.
+	 *             If the exception thrown is an <code>InvocationTargetException</code>, extract its
+	 *             <code>cause</code> and pass it to the <code>ELException</code> constructor.
+	 * @since 2.2
+	 */
+	@Override
+	public Object invoke(ELContext context, Object base, Object method, Class<?>[] paramTypes, Object[] params) {
+		if (context == null) {
+			throw new NullPointerException();
+		}
+		Object result = null;
+		if (isResolvable(base)) {
+			if (params == null) {
+				params = new Object[0];
+			}
+			String name = method.toString();
+			Method target = findMethod(base, name, paramTypes, params.length);
+			if (target == null) {
+				throw new MethodNotFoundException("Cannot find method " + name + " with " + params.length + " parameters in " + base.getClass());
+			}
+			try {
+				result = target.invoke(base, coerceParams(getExpressionFactory(context), target, params));
+			} catch (InvocationTargetException e) {
+				throw new ELException(e.getCause());
+			} catch (IllegalAccessException e) {
+				throw new ELException(e);
+			}
+			context.setPropertyResolved(true);
+		}
+		return result;
+	};
+
+	private Method findMethod(Object base, String name, Class<?>[] types, int paramCount) {
+		if (types != null) {
+			try {
+				return findAccessibleMethod(base.getClass().getMethod(name, types));
+			} catch (NoSuchMethodException e) {
+				return null;
+			}
+		}
+		Method varArgsMethod = null;
+		for (Method method : base.getClass().getMethods()) {
+			if (method.getName().equals(name)) {
+				int formalParamCount = method.getParameterTypes().length;
+				if (method.isVarArgs() && paramCount >= formalParamCount - 1) {
+					varArgsMethod = method;
+				} else if (paramCount == formalParamCount) {
+					return findAccessibleMethod(method);
+				}
+			}
+		}
+		return varArgsMethod == null ? null : findAccessibleMethod(varArgsMethod);
+	}
+
+	/**
+	 * Lookup an expression factory used to coerce method parameters in context under key
+	 * <code>"javax.el.ExpressionFactory"</code>.
+	 * If no expression factory can be found under that key, use a default instance created with
+	 * {@link ExpressionFactory#newInstance()}.
+	 * @param context
+	 *            The context of this evaluation.
+	 * @return expression factory instance
+	 */
+	private ExpressionFactory getExpressionFactory(ELContext context) {
+		Object obj = context.getContext(ExpressionFactory.class);
+		if (obj instanceof ExpressionFactory) {
+			return (ExpressionFactory)obj;
+		}
+		if (defaultFactory == null) {
+			defaultFactory = ExpressionFactory.newInstance();
+		}
+		return defaultFactory;
+	}
+	
+	private Object[] coerceParams(ExpressionFactory factory, Method method, Object[] params) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.